Generate "Is top document in fullscreen?" for site isolation

beidson opened the pull request:
#### 0bca9693851541aa0fa092fed6cb2362a6ea0c75
<pre>
Generate &quot;Is top document in fullscreen?&quot; for site isolation
<a href="https://bugs.webkit.org/show_bug.cgi?id=285556">https://bugs.webkit.org/show_bug.cgi?id=285556</a>
<a href="https://rdar.apple.com/problem/142504863">rdar://problem/142504863</a>

Reviewed by Alex Christensen.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::populateDocumentSyncDataForNewlyConstructedDocument):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::willEnterFullscreen):
(WebCore::FullscreenManager::didExitFullscreen):
(WebCore::FullscreenManager::updatePageFullscreenStatusIfTopDocument):
(WebCore::FullscreenManager::clear):
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setTopDocumentHasFullscreenElement):
(WebCore::Page::topDocumentHasFullscreenElement):
(WebCore::Page::updateProcessSyncData):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/ProcessSyncData.in:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::requestExitFullScreen):

Canonical link: <a href="https://commits.webkit.org/288737@main">https://commits.webkit.org/288737@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6eb4bb772010268ee78a1fd25ec9db4f1e1cb896

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84259 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3880 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38563 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89335 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35268 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86344 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3968 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11855 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65533 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23369 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87305 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2978 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76563 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45823 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2928 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30793 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34317 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73831 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31558 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90717 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11526 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8368 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73985 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11752 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72388 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73187 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18107 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17507 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15950 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2891 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11478 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16954 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11327 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14803 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13100 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -503,7 +504,9 @@ bool FullscreenManager::willEnterFullscreen(Element& element, HTMLMediaElementEn
     m_pendingFullscreenElement = nullptr;
+
     m_fullscreenElement = &element;
+    protectedPage->setTopDocumentHasFullscreenElement(true);
```
I think all Documents own a FullscreenManager.  I'd expect to see a main frame check before making this call

achristensen07 commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -604,8 +607,7 @@ bool FullscreenManager::didExitFullscreen()
 
-    m_fullscreenElement = nullptr;
-    m_pendingFullscreenElement = nullptr;
+    clear();
```
This adds a clearing of m_pendingPromise that wasn't there before.  Is that intentional, or does that matter?

beidson replied:
Intentional, but on further review it does matter. Fixing.

achristensen07 commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -715,6 +717,9 @@ void FullscreenManager::setAnimatingFullscreen(bool flag)
 void FullscreenManager::clear()
 {
+    if (RefPtr protectedPage = page())
+        protectedPage->setTopDocumentHasFullscreenElement(false);
```
Ditto about main frame check.

achristensen07 commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -503,7 +504,10 @@ bool FullscreenManager::willEnterFullscreen(Element& element, HTMLMediaElementEn
+
     m_fullscreenElement = &element;
+    updatePageFullscreenStatusIfTopDocument();
+    protectedPage->setTopDocumentHasFullscreenElement(true);
```
I still think this line needs to be removed.  It was replaced by the previous line which is more correct.

beidson committed:
Generate "Is top document in fullscreen?" for site isolation
https://bugs.webkit.org/show_bug.cgi?id=285556
rdar://problem/142504863

Reviewed by Alex Christensen.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::populateDocumentSyncDataForNewlyConstructedDocument):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::willEnterFullscreen):
(WebCore::FullscreenManager::didExitFullscreen):
(WebCore::FullscreenManager::updatePageFullscreenStatusIfTopDocument):
(WebCore::FullscreenManager::clear):
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setTopDocumentHasFullscreenElement):
(WebCore::Page::topDocumentHasFullscreenElement):
(WebCore::Page::updateProcessSyncData):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/ProcessSyncData.in:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::requestExitFullScreen):

Canonical link: https://commits.webkit.org/288737@main

nt1m commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -713,11 +718,26 @@ void FullscreenManager::setAnimatingFullscreen(bool flag)
+    if (!protectedPage)
+        return;
+
+    protectedPage->setTopDocumentHasFullscreenElement(m_fullscreenElement);
```
This is a behavior change from before. Potentially an escape.

`m_fullscreenElement` does not track the case where the top document is implicitly fullscreened by its child frames but only when top document is explicitly fullscreened.

`fullscreenElement()` would have been the right thing to use. 

Also, this changeset doesn't appear to broadcast the fullscreen status in the implicit case.
